shell/common/persistent_cache.cc

```diff
@@ -0,0 +1,111 @@
+// Copyright 2018 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/common/persistent_cache.h"
+
+#include <memory>
+#include <string>
+
+#include "flutter/fml/file.h"
+#include "flutter/fml/mapping.h"
+#include "flutter/fml/paths.h"
+#include "flutter/fml/trace_event.h"
+#include "flutter/shell/version/version.h"
+#include "third_party/skia/include/utils/SkBase64.h"
+
+namespace shell {
+
+static std::string SkKeyToFilePath(const SkData& data) {
+  if (data.data() == nullptr || data.size() == 0) {
+    return "";
+  }
+
+  const size_t encoded_length =
+      SkBase64::Encode(data.data(), data.size(), nullptr);
+
+  if (encoded_length == 0) {
+    return "";
+  }
+
+  auto encoded_bytes = SkData::MakeUninitialized(encoded_length);
+
+  if (!encoded_bytes) {
+    return "";
+  }
+
+  SkBase64::Encode(data.data(), data.size(), encoded_bytes->writable_data());
+
+  return std::string{reinterpret_cast<const char*>(encoded_bytes->data()),
+                     encoded_length};
+}
+
+PersistentCache* PersistentCache::GetCacheForProcess() {
+  static std::unique_ptr<PersistentCache> gPersistentCache;
+  static std::once_flag once = {};
+  std::call_once(once, []() { gPersistentCache.reset(new PersistentCache()); });
+  return gPersistentCache.get();
+}
+
+PersistentCache::PersistentCache()
+    : cache_directory_(CreateDirectory(fml::paths::GetCachesDirectory(),
+                                       {
+                                           "flutter_engine",           //
+                                           GetFlutterEngineVersion(),  //
+                                           "skia",                     //
+                                           GetSkiaVersion()            //
+                                       },
+                                       fml::FilePermission::kReadWrite)) {
+  if (!cache_directory_.is_valid()) {
+    FML_LOG(ERROR) << "Could not acquire the persistent cache directory. "
+                      "Caching of GPU resources on disk is disabled.";
+  }
+}
+
+PersistentCache::~PersistentCache() = default;
+
+// |GrContextOptions::PersistentCache|
+sk_sp<SkData> PersistentCache::load(const SkData& key) {
+  if (!cache_directory_.is_valid()) {
+    return nullptr;
+  }
+
+  std::lock_guard<std::mutex> lock(mutex_);
```

Comment:
What is the concurrency model for this cache?

This lock is going to be held during some long I/O operations.  The lock might not be necessary if the filesystem already provides all of the guarantees that you need (in particular, the use of WriteAtomically should ensure that a reader always sees a complete version of the file at a given path)

Reply:
Right. This was added before the WriteAtomic path. Removed.